src/icelake/icelake_utf8_length_from_utf16.inl.cpp

```diff
@@ -1,84 +1,288 @@
+  pos = size;
+
+  count += pos;
+  return count;
+}
+
+template <endianness big_endian>
+simdutf_really_inline size_t icelake_utf8_length_from_utf16_with_replacement(
+    const char16_t *in, size_t size) {
+  ///////
+  // We repeat 3 times the same algorithm.
+  // First, we proceed with an unrolled loop of 2*N char16 values (for speed).
+  // Second, we process N char16 values.
+  // Finally, we process the remaining char16 values (less than N).
+  ///////
+  using vector_u16 = simd16<uint16_t>;
+  constexpr size_t N = vector_u16::ELEMENTS; // 32 on AVX-512
+  if (N + 1 > size) {
+    return scalar::utf16::utf8_length_from_utf16_with_replacement<big_endian>(
+        in, size);
+  } // special case for short inputs
+  size_t pos = 0;
+
+  const __m512i byteflip = _mm512_setr_epi64(
+      0x0607040502030001, 0x0e0f0c0d0a0b0809, 0x0607040502030001,
+      0x0e0f0c0d0a0b0809, 0x0607040502030001, 0x0e0f0c0d0a0b0809,
+      0x0607040502030001, 0x0e0f0c0d0a0b0809);
+
+  size_t count = 0;
+  size_t mismatched_count = 0;
+
+  if (scalar::utf16::is_low_surrogate<big_endian>(in[0])) {
+    mismatched_count += 1;
+  }
+
+  for (; pos < (size - 1) / (2 * N) * (2 * N); pos += 2 * N) {
+    __m512i current1 =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos));
+    if (!match_system(big_endian)) {
+      current1 = _mm512_shuffle_epi8(current1, byteflip);
+    }
+    __m512i next1 =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos + 1));
+    if (!match_system(big_endian)) {
+      next1 = _mm512_shuffle_epi8(next1, byteflip);
+    }
+    __m512i current2 =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos + N));
+    if (!match_system(big_endian)) {
+      current2 = _mm512_shuffle_epi8(current2, byteflip);
     }
+    __m512i next2 =
+        _mm512_loadu_si512(reinterpret_cast<const __m512i *>(in + pos + N + 1));
+    if (!match_system(big_endian)) {
+      next2 = _mm512_shuffle_epi8(next2, byteflip);
+    }
+
+    __m512i lb_masked1 =
+        _mm512_and_si512(current1, _mm512_set1_epi16(uint16_t(0xfc00)));
+    __m512i block_masked1 =
+        _mm512_and_si512(next1, _mm512_set1_epi16(uint16_t(0xfc00)));
+    __mmask32 lb_is_high1 = _mm512_cmpeq_epi16_mask(
+        lb_masked1, _mm512_set1_epi16(uint16_t(0xd800)));
+    __mmask32 block_is_low1 = _mm512_cmpeq_epi16_mask(
+        block_masked1, _mm512_set1_epi16(uint16_t(0xdc00)));
+    __mmask32 illseq1 = _kxor_mask32(lb_is_high1, block_is_low1);
+    __mmask32 is_surrogate1 = _mm512_cmpeq_epi16_mask(
+        _mm512_and_si512(current1, _mm512_set1_epi16(uint16_t(0xf800))),
+        _mm512_set1_epi16(uint16_t(0xd800)));
+    __mmask32 c01 = _mm512_cmpneq_epi16_mask(
+        _mm512_and_si512(current1, _mm512_set1_epi16(uint16_t(0xff80))),
+        _mm512_setzero_si512());
+    __mmask32 c11 = _mm512_cmpneq_epi16_mask(
+        _mm512_and_si512(current1, _mm512_set1_epi16(uint16_t(0xf800))),
+        _mm512_setzero_si512());
```

Comment:
175 to 190 can benefit from _mm512_test_epi16_mask as well

Reply:
there are couple of other similar optimizations in this file as well

Reply:
You are correct. I missed a few more.